overlord/snapstate/snapstate.go

```diff
@@ -2102,7 +2105,32 @@ func splitRefresh(essential, nonEssential []minimalInstallInfo, updateFunc func(
 			}
 
 			appStartTask.WaitFor(baseEndTask)
+			crossSetDependency = true
+		}
+	}
+
+	// essential snaps don't use epochs at the moment so we only need to consider
+	// re-refreshes for the non-essential refreshes
+	if len(nonEssentialTss.Refresh) > 0 && !flags.NoReRefresh {
+		rerefreshSetup := reRefreshSetup{
+			UserID: userID,
+			Flags:  flags,
+		}
+
+		// if there are no cross-set dependencies, the rerefresh can be done
+		// before the reboot. If some app does need to wait for the reboot, the
+		// rerefresh check needs to wait for it so we do it at the end as usual
+		if !crossSetDependency {
+			for _, ts := range nonEssentialTss.Refresh {
+				for _, t := range ts.Tasks() {
+					rerefreshSetup.TaskIDs = append(rerefreshSetup.TaskIDs, t.ID())
+				}
+			}
 		}
+
+		rerefresh := st.NewTask("check-rerefresh", reRefreshSummary(nonEssentialUpdated, flags))
+		rerefresh.Set("rerefresh-setup", rerefreshSetup)
+		nonEssentialTss.Refresh = append(nonEssentialTss.Refresh, state.NewTaskSet(rerefresh))
```

Comment:
it would be good if this was also finalizeUpdate. possibly passing some extra optional args

Reply:
Added the option to pass tasks to be considered into `finalizeUpdate`